test/integration/distributed/setup/node_test.go

```diff
@@ -86,4 +92,64 @@ var _ = Describe("Node registration", func() {
 			return helpers.ListKeys(etcdEndpoint, fmt.Sprintf("/%s/nodes/%s:%d", namespace, nodeHost, ports[0]))
 		}, flags.EventuallyTimeout).Should(BeNil())
 	})
+
+	It("should register multiple data nodes in the node selector successfully", func() {
+		namespace := "data-test"
+		nodeDataHosts := []string{"data-node-1", "data-node-2"}
+		nodeDataIDs := make([]string, len(nodeDataHosts))
+		ports, err := test.AllocateFreePorts(2)
+		Expect(err).NotTo(HaveOccurred())
+
+		var closeFns []func()
+		for i, nodeDataHost := range nodeDataHosts {
+			addr := fmt.Sprintf("%s:%d", host, ports[i])
+			closeFns = append(closeFns, setup.CMD("data",
+				"--namespace", namespace,
+				"--grpc-host="+host,
+				fmt.Sprintf("--grpc-port=%d", ports[i]),
+				"--etcd-endpoints", etcdEndpoint,
+				"--node-host-provider", "flag",
+				"--node-host", nodeDataHost))
+			nodeDataIDs[i] = fmt.Sprintf("%s:%d", nodeDataHost, ports[i])
+			Eventually(
+				helpers.HealthCheck(addr, 10*time.Second, 10*time.Second, grpc.WithTransportCredentials(insecure.NewCredentials())),
+				flags.EventuallyTimeout).Should(Succeed())
+			Eventually(func() (map[string]*databasev1.Node, error) {
+				return helpers.ListKeys(etcdEndpoint, fmt.Sprintf("/%s/nodes/%s:%d", namespace, nodeDataHost, ports[i]))
+			}, flags.EventuallyTimeout).Should(HaveLen(1))
+		}
+
+		// TODO: refactor the following startup logic
```

Comment:
Any idea how we can better organize such (fine-grained) services setup

Reply:
Would you be able to use a mock to test the behavior of the selector? 

I have introduced a cluster integration in #333 . We can use it to test the cluster selector.